providers/class-two-factor-backup-codes.php

```diff
@@ -177,15 +177,23 @@ public function user_options( $user ) {
 							// Update counter.
 							$( '.two-factor-backup-codes-count' ).html( response.data.i18n.count );
 
-							// Build the download link.
-							var txt_data = 'data:application/text;charset=utf-8,' + '\n';
-							txt_data += response.data.i18n.title.replace( /%s/g, document.domain ) + '\n\n';
+							// Build the downloaded file contents.
+							var txt_data = response.data.i18n.title.replace( /%s/g, document.domain ) + '\n\n';
```

Comment:
Could we offload this to PHP and make it part of the AJAX response along with `codes`?

https://github.com/WordPress/two-factor/blob/5363ca67fd6121f1a8ae9fea6e266f97782693f7/providers/class-two-factor-backup-codes.php#L241-L261

This would allow us to make that helper testable and avoid all the browser encoding/escaping nuances.

Reply:
I think that's a reasonable suggestion.

I think it'd be reasonable to shift the generation of the `data:` uri to PHP as well, and just return that along with the XHR payload, removing most of the JS reliance.